### PR TITLE
Feature/payments plugin paypal

### PR DESCRIPTION
**Pull request by sonntag-philipp**
# Description

Please include a summary of the changes and the related issue.

# Breaking changes

Does this PR include any breaking changes we should be aware of?

# Screenshots

You can add screenshots here if applicable.

# Checklist

📌 Always:
- [ ] I have set a clear title
- [ ] My PR is small and contains a single feature
- [ ] I have [checked my own PR](## "Fix typo's and remove unused or commented out code")

👍 Most of the time:
- [ ] I have added or updated test cases
- [ ] I have updated the README if needed

### CODE REVIEW

File: packages/payments-plugin/package.json [hunk 1]
```diff
@@ -29,7 +30,8 @@
         "stripe": "13.x"
     },
     "dependencies": {
-        "currency.js": "2.0.4"
+        "currency.js": "2.0.4",
+        "msw": "^2.4.4"
```
Comment (sonntag-philipp) [inline packages/payments-plugin/package.json]: I can not use nock with the native fetch api that is used within the paypal integration.
See this issue for more information: https://github.com/nock/nock/issues/2397

The fetch Api was introduced in NodeJS v18, so there shouldn't be any compatibility problems I think.

Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: See https://www.npmjs.com/package/msw for more information

Comment (mikicho) [reply packages/payments-plugin/package.json]: `nock` does support the native `fetch` in the beta version `npm I nock@beta.` It's stable and has been in beta for a few months now.

Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: Thank you, I will try it again with the beta release and change it for our tests if necessary.

File: .vscode/launch.json
Comment (sonntag-philipp) [inline .vscode/launch.json:1]: I'm not quite sure if most of the maintainers use WebStorm, but this makes it way easier to execute a specific test file in VSCode. 

Also had to make some changes to the `get-package-dir` to make it work not only for the e2e within the `payments-plugin`.

File: packages/payments-plugin/e2e/fixtures/paypal.fixtures.ts
Comment (sonntag-philipp) [inline packages/payments-plugin/e2e/fixtures/paypal.fixtures.ts:1]: Mockdata for the mocked PayPal REST Api.

File: packages/payments-plugin/e2e/paypal-dev-server.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import { setShipping } from './payment-helpers';
+import { PayPalSdkPlugin } from './paypal-sdk.plugin';
+
+/**
+ * This should only be used to locally test the PayPal payment plugin
+ * Make sure you have `PAYPAL_CLIENT_ID=xxxx`, `PAYPAL_CLIENT_SECRET=xxxx`
+ * and `PAYPAL_MERCHANT_ID=xxxx` in your .env file
+ */
+/* eslint-disable @typescript-eslint/no-floating-promises */
+async function runPayPalDevServer() {
```
Comment (sonntag-philipp) [inline packages/payments-plugin/e2e/paypal-dev-server.ts:25]: This is almost fully copied from the Mollie dev server.

File: packages/payments-plugin/e2e/paypal-dev-server.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    });
+    await setShipping(shopClient);
+    await shopClient.query(APPLY_COUPON_CODE, { couponCode: 'FREE_SHIPPING' });
+
+    await shopClient.query(TRANSITION_TO_STATE, {
+        state: 'ArrangingPayment',
+    });
+
+    // eslint-disable-next-line no-console
+    console.log('Auth Token: ', shopClient.getAuthToken());
```
Comment (sonntag-philipp) [inline packages/payments-plugin/e2e/paypal-dev-server.ts:111]: This token is printed, so we can enter it in the Web UI of the paypal plugin. I would be open to setting this automatically in the Vendure REST Endpoint, but I wasn't quite sure how.

File: packages/payments-plugin/e2e/paypal-sdk.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import { ADD_PAYMENT_TO_ORDER, CREATE_PAYPAL_ORDER } from './payment-helpers';
+
+const PAYPAL_SDK_PAYMENT_PLUGIN_OPTIONS = Symbol('PAYPAL_SDK_PAYMENT_PLUGIN_OPTIONS');
+export interface PayPalSdkPluginOptions {
+    apiUrl: string;
+    clientId: string;
+}
+
+// @ts-ignore
+function html(strings, ...values) {
```
Comment (sonntag-philipp) [inline packages/payments-plugin/e2e/paypal-sdk.plugin.ts:19]: Well. I know we shouldn't integrate code only for syntax highlighting, but as this isn't code of the library itself and only exists for testing purposes, one could argue that it's still okay. 
Especially as it doesn't require additional dependencies.

File: packages/payments-plugin/src/paypal/paypal-authorization.service.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                },
+            });
+
+            if (!response.ok) {
+                throw new Error('Failed to get PayPal order details.');
+            }
+
+            return (await response.json()) as PayPalOrderInformation;
+        } catch (error) {
+            throw new InternalServerError('PayPal integration failed');
```
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/paypal-authorization.service.ts]: Make better error logging within the catch blocks of the PayPal services.

File: packages/payments-plugin/src/paypal/paypal-authorization.service.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        ctx: RequestContext,
+        authorizationId: string,
+    ): Promise<PayPalOrderInformation> {
+        const accessToken = (await this.authenticate(ctx)).access_token;
+
+        const response = await fetch(
+            `${this.options.apiUrl}/v2/payments/authorizations/${authorizationId}/capture`,
+            {
+                method: 'POST',
+                headers: {
```
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/paypal-authorization.service.ts]: Headers could also be composed within the base class.

File: packages/payments-plugin/src/paypal/paypal-base.service.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        if (!paymentMethod.handler.args) {
+            throw new InternalServerError('No args defined for PayPal payment method');
+        }
+
+        return paymentMethod.handler.args;
+    }
+
+    protected async getPaymentMethod(ctx: RequestContext): Promise<PaymentMethod | undefined> {
+        const allPaymentMethods = await this.paymentMethodService.findAll(ctx);
+        const paymentMethod = allPaymentMethods.items.find(item => item.handler?.code === handlerCode);
```
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/paypal-base.service.ts]: Return directly.

File: packages/payments-plugin/src/paypal/paypal-order.service.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        } catch (error: any) {
+            const message = 'Failed to get PayPal order details';
+            if (error instanceof Response) {
+                const responseClone = error.clone();
+                Logger.error(message, loggerCtx, await responseClone.text());
+            }
+
+            Logger.error(message, loggerCtx, error?.stack);
+            // Throw a generic error to avoid leaking sensitive information
+            throw new InternalServerError(message);
```
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/paypal-order.service.ts:108]: Use more specific error messages instead of InternalServerError

File: packages/payments-plugin/src/paypal/types.ts
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/types.ts:1]: Re-validate if PayPal NPM Package types are a valid option for the typings.

**Comment by sonntag-philipp** [comment]
I still have to implement the idempotency headers supported by the PayPal Api, even though it already works without them. This could also be done is a future PR.

File: packages/payments-plugin/package.json (hunk 1 above)
Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: Switched to nock and updated to beta version.

File: packages/payments-plugin/e2e/paypal-dev-server.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+async function runPayPalDevServer() {
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    require('dotenv').config();
+
+    const { PAYPAL_CLIENT_ID, PAYPAL_CLIENT_SECRET, PAYPAL_MERCHANT_ID } = process.env;
+    if (!PAYPAL_CLIENT_ID || !PAYPAL_CLIENT_SECRET || !PAYPAL_MERCHANT_ID) {
+        throw new Error('Please provide PayPal credentials in the .env file');
+    }
+
+    const apiUrl = 'https://api-m.sandbox.paypal.com';
```
Comment (DanielBiegler) [inline packages/payments-plugin/e2e/paypal-dev-server.ts]: I know its probably unlikely for paypal to change their sandbox URL but it could prove useful to get this URL also from the .env with the current value as default.

File: packages/payments-plugin/src/paypal/paypal.shop-resolver.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    @Allow(Permission.Owner)
+    async createPayPalOrder(@Ctx() ctx: RequestContext): Promise<PayPalOrder | undefined> {
+        const sessionOrder = await this.activeOrderService.getActiveOrder(ctx, {});
+
+        if (!sessionOrder) {
+            throw new IllegalOperationError('Session has no active order');
+        }
+        if (sessionOrder.state !== 'ArrangingPayment') {
+            throw new IllegalOperationError('Order must be in arranging payment state');
+        }
```
Comment (DanielBiegler) [inline packages/payments-plugin/src/paypal/paypal.shop-resolver.ts:33]: I think these could be internationalized with existing error message for example there are:

```
    "NO_ACTIVE_ORDER_ERROR": "There is no active Order associated with the current session",
```

and

```
    "ORDER_MODIFICATION_ERROR": "Order contents may only be modified when in the \"AddingItems\" state",
    "ORDER_PAYMENT_STATE_ERROR": "A Payment may only be added when Order is in \"ArrangingPayment\" state",
    "ORDER_STATE_TRANSITION_ERROR": "Cannot transition Order from \"{ fromState }\" to \"{ toState }\"",
```

Might be good to take a second pass over all thrown errors and see what can be internationalized and reused.

Comment (sonntag-philipp) [reply packages/payments-plugin/src/paypal/paypal.shop-resolver.ts:33]: Really good feeback, thank you :) Will definetly do that!

**Comment by sonntag-philipp** [comment]
Renamed the commit messages to fit other scopes and rebased on minor branch.

File: packages/payments-plugin/package.json (hunk 1 above)
Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: Updating the nock version broke other test cases. @mikicho maybe you can take a brief look at the test case in hopes that you see the issue directly: https://github.com/vendure-ecommerce/vendure/blob/dcf7b0a023532e815b7aff0995389db4f9ef6bb7/packages/payments-plugin/e2e/stripe-payment.e2e-spec.ts#L135

I will try to locate the issue myself, just thought you may see the issue directly :)

Comment (mikicho) [reply packages/payments-plugin/package.json]: Seems ok to me.. what error do you get?

Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: I could not get an error for now, it just hits the timeout of the test case. But I found the exact location where the timeout happens, which is here: 
https://github.com/vendure-ecommerce/vendure/blob/dcf7b0a023532e815b7aff0995389db4f9ef6bb7/packages/payments-plugin/src/stripe/stripe.service.ts#L173

Comment (sonntag-philipp) [reply packages/payments-plugin/package.json]: To be more precise here: 
https://github.com/stripe/stripe-node/blob/768206df1b9d025853d5e34e6046beb502effdd9/src/net/NodeHttpClient.ts#L83

The callback of `req.on('response')` won't be called after upgrading to `nock@14.0.0-beta.15` and with `stripe@17.2.1`.

See this issue: https://github.com/nock/nock/issues/2785

**Comment by sonntag-philipp** [comment]
Hey @michaelbromley thank you for the work you have done for this great FOSS project.

I just wanted to ask if this has any chance of getting merged in the future. I still have to add documentation, will do that after a review. I don't want to rush you, just want to know if it's even possible that this gets merged :D

**Comment by taxilian** [comment]
If this isn't accepted you should definitely release it as a standalone plugin – I've been needing this.

**Comment by Lennix** [comment]
@michaelbromley We are also interested in this getting merged. ✅

**Comment by michaelbromley** [comment]
Hi @sonntag-philipp!

Apologies for the slow response to this one. And wow, this is an epic effort you have put in here!

In principle I think it does make sense to have a PayPal integration in the core plugins, since it is just as prominent as Stripe, Mollie and Braintree.

My only reservation is this: the payment plugins are the hardest for me & the Vendure team to maintain because in order to really, deeply understand and maintain it, you need to _use_ it. For instance, with the Mollie plugin I rely on Martijn from Pinelab, who uses it extensively and I trust him to just keep it working.

Braintree is used by me on other projects.

Stripe is somewhat problematic - recently we had a PR that I approved, but it turned out to break things for other users. The reason is that I have neither the time nor desire to do what it takes to properly maintain these: fully understanding the APIs, setting up test accounts, setting up front-end integrations to test the full flows, keeping up-to-date with API changes, etc.

So basically I would consider putting Paypal in core if I have some confidence level that someone is willing to be the primary maintainer. Is that something you think could work for you?

**Comment by sonntag-philipp** [comment]
Thank you all for the feedback! 

Hi @michaelbromley!

I would really like to get this plugin into core. As you already stated it was much effort and I tried to make it as robust as possible with a narrow feature scope that I would be able to support.

Even though I kept this plugin intentionally small (e.g. with just a 2 step flow and no 1 step flow), I am not quite sure how much effort it can be to maintain a small part within a bigger open source project. This is my first bigger contribution to FOSS.

Maybe you can give me some estimations where my responsibilities would be as a primary maintainer? Especially new feature requests would be hard to work on as I don't have much time for that, whereas bug fixes and security issues are really interesting to me. Especially as that would harden our own project as well.

Just one point to consider is that we haven't used this plugin in any production environment yet.

You can also contact me via Discord if you want to. _flypsi_ is my name on the Vendure Discord 😄

**Comment by taxilian** [comment]
I would absolutely be willing to help with testing this and maintaining a longer term, since this is something that I will need… That said, I will not always be able to be responsive quickly, and so I would not be a good primary maintainer. In my experience, the main challenge with being the primary maintainer for a piece of a larger project like this is the need to preemptively test things as new releases come out and particularly to test and approve any poor requests related to this feature since the other maintainers will not have an easy way of testing it

**Comment by michaelbromley** [comment]
Thanks @sonntag-philipp  & @taxilian 

So the main responsibility is indeed that bugs get fixed and (especially) any security issues are swiftly dealt with. 

In terms of features, in my experience so far once the integration is done, there's not that much more requested in terms of features. The main things that came up are making sure it works in a multi-channel environment (this is currently an ongoing topic in the Stripe plugin, for instance).

Then of course, if anything changes with Paypal's APIs that would require updates here, you'd need to be on top of that. Usually payment providers give lots of upfront notice on such things though.

@dlhck do you have any input on this? I think PayPal is worth having in the payments package as long as Phillip and others can maintain it.

**Comment by dlhck** [comment]
@michaelbromley I think having this community maintained makes sense.

**Comment by sonntag-philipp** [comment]
Thanks @taxilian for the offer. I'm really happy that this feature found supporters 😄

With all this in mind I am comfortable with being the primary maintainer of the PayPal integration, as we need this integration for work projects anway, so I will be working on it one way or another.

I am going to add some documentation to this plugin in the coming days.

File: packages/payments-plugin/src/paypal/paypal-authorization.service.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                },
+            });
+
+            if (!response.ok) {
+                throw response;
+            }
+
+            return (await response.json()) as PayPalOrderInformation;
+        } catch (error: any) {
+            const message = 'error.paypal-authorization-details-failed';
```
Comment (sonntag-philipp) [inline packages/payments-plugin/src/paypal/paypal-authorization.service.ts:36]: Even though SonarQube already mentioned this code is often duplicated, I want to make sure that errors are correctly handled this way in Vendure. Maybe you guys have some feedback on this :)

**Comment by sonntag-philipp** [comment]
I've added a small documentation on how to get started using this plugin. Reviews are welcome right now :)

@michaelbromley @taxilian

**Comment by dfernandesbsolus** [comment]
@michaelbromley @sonntag-philipp Hey guys, is this pr outdated? This plugin would be a great complement to Braintree which is not available in many countries.

**Comment by michaelbromley** [comment]
Hi @sonntag-philipp - apologies for the long time with no response on this. We'd be interested in getting this merged for the next minor release, v3.4. The conflicts look rather minor, would you have a bit of time to get that resolved?

If not, let me know and one of us can take that on.

**Comment by sonntag-philipp** [comment]
@michaelbromley Oh, I thought the PR was "dismissed" (or something like that) so I moved the PayPal package to my local workspace (without relevant changes) :D

I don't have much time until at least Monday, so I could fix the conflicts until Tuesday next week. If you need it more early for the 3.4 minor release, feel free to work on that :)

**Comment by michaelbromley** [comment]
@sonntag-philipp there's no rush on this. We're looking around the end of June for the 3.4 release.

No, the PR was not dismissed - we just dropped the ball on the communication front with all the things we have going on.

**Comment by sonntag-philipp** [comment]
Sorry guys, I have to meet a deadline on another project. This PR has to wait some more time.

Just the merge would not be the problem, but I think I do not I will be able to respond somewhat fast on issues that occur when this plugin is used.

At least until the end of july I will not be able to spend much time on this.

@michaelbromley

File: packages/payments-plugin/src/paypal/paypal.plugin.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        PayPalCaptureService,
+    ],
+    configuration: config => {
+        config.paymentOptions.paymentMethodHandlers.push(paypalPaymentMethodHandler);
+
+        return config;
+    },
+    compatibility: '^3.0.0',
+    adminApiExtensions: {
+        schema: shopSchemaExtensions,
```
Comment (vjung28) [inline packages/payments-plugin/src/paypal/paypal.plugin.ts:139]: isn't the shop and admin schema swapped here?

**Review by vjung28 — CHANGES_REQUESTED** [review]

File: packages/payments-plugin/src/paypal/paypal.plugin.ts (hunk 12 above)
Comment (dlhck) [reply packages/payments-plugin/src/paypal/paypal.plugin.ts:139]: true! good catch

**Comment by dlhck** [comment]
@sonntag-philipp I would really like to get this PR merged as it is a great improvement.  Do you know if you can spend some more time to finish this? Please let me know.

**Comment by larsdecker** [comment]
Any news on this topic? It would be really good to see a Paypal Integration in Vendure.

**Comment by michaelbromley** [comment]
Hey @sonntag-philipp - as you might have seen we have this new repo for community integrations:https://github.com/vendurehq/community-plugins

If you still have the time and willingness to push this through, we'd be very open to a PR to that repo. Apologies that this PR went for so long without resolution - that's one of the reasons we split this kind of thing out into the separate repo: it allows us to move faster with changes like this.

I'm going to close this PR now because next week we're fully removing all the payment plugins from this repo anyway. This does _not_ mean we aren't interested in your work, if you're willing to contribute 👍
